Enable fixed fast_mode for complex

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56964 Clean up naming and comments
* #55871 Refactor fast gradcheck
* **#55699 Enable fixed fast_mode for complex**
* #55692 Refactor C->C to C->R twice

Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component

Differential Revision: [D28007887](https://our.internmc.facebook.com/intern/diff/D28007887)

soulitzer committed:
Enable fixed fast_mode for complex

[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -314,9 +314,10 @@ def get_jvp_wrt_specific_input(fn, input_idx, inputs, outputs, u, eps) -> List[t
+    if ur.layout != torch.sparse_coo:
+        ur = ur.reshape(input_to_perturb.shape)
+        ui = ui.reshape(input_to_perturb.shape)
+    return compute_numerical_jacobian_cols(jvp_fn, ur * eps, input.is_complex(), ui * eps)
```
nit: ordering of the arguments looks a bit weird here. Maybe use `delta_i=ui * eps` to make it less weird to have the `input.is_complex()` in between the two.

soulitzer replied:
I've updated to use only a single delta, see other comment above

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -867,36 +870,46 @@ def fast_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
     g_cpu = torch.Generator()
-    all_u = [vec_from_tensor(inp, g_cpu) for inp in inp_tensors]
+    all_ur = [vec_from_tensor(inp, g_cpu, True) for inp in inp_tensors]
+    all_ui = [vec_from_tensor(inp, g_cpu, True) for inp in inp_tensors]
```
Does this mean that you always generate the ui, even for real inputs and just end up never using it?

soulitzer replied:
Updated to only generate when necessary.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -867,36 +870,46 @@ def fast_gradcheck(func, func_out, tupled_inputs, outputs, eps, rtol,
+                ai = av.select(-1, 1)
+                all_analytical[i].append(ar.dot(ur) + 1j * ai.dot(ui))
+            else:
+                all_analytical[i].append(dot_with_type_promotion(a.T.squeeze(0), ur))
```
why do you need dot_with_type_promotion for the real case but it wasn't needed before?

soulitzer replied:
This is not needed actually, this was resolved by the "complex downcasting" suggestion above

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -790,7 +792,7 @@ def allclose_with_type_promotion(a, b, rtol, atol):
 
 
-def vec_from_tensor(x, generator):
+def vec_from_tensor(x, generator, always_float64=False):
```
nit: what you want here is not reall a "always_float64" but just always downcast complex to their element type.
Maybe a simple if statement here that transforms complex64 -> float32 and complex128 -> float64 would be enough?

albanD reviewed (commented):
The logic looks ok for me.

I wonder if there would be a way to have `u` as an object so that you don't need to add a new argument to every function.
That object would contain one or two real valued vectors.
And you unpack it properly depending if you're considering complex or not when you need it. What do you think?

soulitzer commented on torch/autograd/gradcheck.py:
```diff
@@ -201,7 +201,7 @@ def compute_numerical_jacobian_cols(jvp_fn, delta, input_is_complex) -> List[tor
 
     if input_is_complex:  # C -> R
-        ds_dy_tup = jvp_fn(delta * 1j)
+        ds_dy_tup = jvp_fn(delta * 1j) if delta_i is None else jvp_fn(delta_i * 1j)
```
Slow version does not use delta_i. But now I've changed it to use a single parameter delta object which may be either tuple or tensor or python number. It will be always python number in the slow case, and tuple or tensor in the fast case depending on whether that input is complex.

soulitzer commented:
Having a single object `u` instead of having to pass `ui`, `ur` as separate arguments does seem to simply the code a bit, especially once you consider adding more logic to only generate `ui` if the particular input is complex.

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -305,6 +305,23 @@ def jvp_fn(delta):
+    else:
+        if u.layout != torch.sparse_coo:
+            return u.reshape(shape)
+    return u
```
nit: add a comment here saying that we don't need to reshape for sparse Tensors.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -815,6 +843,7 @@ def adjusted_atol(atol, u, v):
     #
     # We see that atol needs to be scaled by v^T M u (where M is an all-ones M x N matrix):
     # v^T M u = \sum_{i} \sum_{j} u_i * v_j = (\sum_{i} u_i)(\sum_{i} v_i)
+    u = u[0] if isinstance(u, tuple) else u
```
Maybe a comment explaining why we ignore the second part of u? Or a todo if that needs to be done later.

soulitzer replied:
Yeah still need to think about how to handle the complex case here. I've moved the other TODO you mentioned to here.

albanD commented on torch/autograd/gradcheck.py:
```diff
@@ -858,44 +887,74 @@ def run_slow_mode_and_get_error(func, tupled_inputs, outputs, input_idx, output_
         for j, n in enumerate(all_numerical_for_input_i):
             a = all_analytical[j][i]
             n = n.to(device=a.device)
+            # TODO: Update adjusted atol
```
This has been done right?

albanD reviewed (approved):
LGTM
Just a couple comments to update and it is good!

soulitzer committed:
Update on "Enable fixed fast_mode for complex"


Todo:
- error message should be updated to say whether the failure is for fn's real or imaginary component


[ghstack-poisoned]
